### DISCUSSION TITLE
can you extend the encrypt middleware example?

### DISCUSSION THREAD

**Discussion by nielsnl68** [ujjwalguptaofficial/JsStore#235] [Q&A]
Hi 

i'm trying out the encrypt middleware and im strucling to understand what should be done inside the `encryptData` and the `decryptData`. Can you show some simple code that would could be inside those functions?

thanks

**Comment by ujjwalguptaofficial** [comment]
check it out example - https://github.com/ujjwalguptaofficial/jsstore-encrypt/tree/main/examples/typescript

particularly, check `initJsStore` method - https://github.com/ujjwalguptaofficial/jsstore-encrypt/blob/00ee9ec58a230712be93e7ed15694e371389d280/examples/typescript/src/storage_service/idb_helper.ts#L61

also check this file which `jsstore-encrypt` expects inside - https://github.com/ujjwalguptaofficial/jsstore-encrypt/blob/main/examples/typescript/static/jsstore-encrypt.js

**Reply by nielsnl68** [reply]
Thanks, i also looked at: 
https://github.com/ujjwalguptaofficial/jsstore-encrypt/blob/main/src/index.ts

 it looks very complicated, i see you are only checking for select and insert. What about "update" is that not needed? And could you please tell a little bit how it works?

I see that JsStoreEncrypt is declared on top of the file but how does it know which object/class belongs to it ?  

Also, telling that the table has encrypted fields with every select and insert/update looks to me a possibility for mistakes by forgetting to add it when you insert a record.

**Reply by ujjwalguptaofficial** [reply]
Its better that i explain with a demo. I will write an article by this weekend.

**Reply by nielsnl68** [reply]
i look forward to it.  Where i'm  most interesting in is the promising part. That is the biggest part i can't follow what happens. 
It is still a javascript future i need to get more under the thumb.

**Reply by ujjwalguptaofficial** [reply]
Hey @nielsnl68 i am still working on some of the major part in jsstore-encrypt. So will take some time to write the article.

I understood after your question that things are little tedious and needs some more changes and explanation.

**Comment by nielsnl68** [comment]
Okey, thanks for your update. 

I saw last weeks changes on the plugin and it was already better understandable.
Not sure of you changed it already but what i saw in the insert/update you create 2 array's one with the promisses and on with the fieldnames. Would it not better readable when you create an async function that takes the fieldname and do the encryption inside and updates the value when done?

Also to speed it a bit up i would suggest you check the fields that needs to be encrypt only once at "init_db" like:

``` javascript
    if (request.name == 'init_db') {
        for (const table of query.tables) {
            if (typeof table.encrypt === "string") { table.encrypt = [table.encrypt]; }
            if (Array.isArray(table.encrypt)) {
                for (const [key, column] of Object.entries(table.columns)) {
                    if (column.encrypt === true) {
                        table.encrypt.push(key);
                    }
                }
                // when there are no fields to be encrypted, then disable the encryption option.
                if (table.encrypt.length===1) table.encrypt = false;
            } else {
                table.encrypt = false;
            }
        }
    }
```
   
in the update/insert section you dan you only need to check of table.encrypt is not false and do a for 1 to length loop to check the fields that need to encrypted. ps, table.encrypt[0] is the key.

good luck.

BTW a little nickpicking please consider renaming the encryptValue into encryptValue**s** and the parameter into value**s** because you try to change more then one value.

**Comment by ujjwalguptaofficial** [comment]
@nielsnl68 have updated code with support for update, where & others cases. I have also updated readme & examples with latest changes. I hope this gives better context now. 

Let me know if you have any suggestion. I will also write article but will take some time, my focus will be still on making examples & readme better that everyone should be able to understand.

**Comment by nielsnl68** [comment]
thanks, good work man. I have on issue to tackel for you. But i will create an other post for that.

**Reply by shaangidwani** [reply]
JSStore Encrypt Data issues

We have implemented JSStore with encrypt but we have some issues in that

When we try to set encrypt date-time time fields it gives the error below :
err message: Invalid array length

When any data is null and that column is set as encrypted then it gives the error:
err message : Cannot read properties of undefined (reading 'encrypt')

Try to select data with decrypt attribute but sometime its gives decrypt data and sometime its gives encrypt data

ex :

getPatient() {
return await this.connection.select({
from: this.tableName,
decrypt: true
})

sometimes Its gives decrypts data but most of the time its gives encrypts data.

I have already submitted issues on jsstore-encrypt repository but I have not gotten any help.

Please anyone who can help on this will be highly appreciated.
